utils/workers/Creators.py

```diff
@@ -0,0 +1,364 @@
+        self.orig_dir = orig_dir
+        self.roi_dir = roi_dir
+        self.output_dir = output_dir
+        self.base_file_name = base_file_name
+        self.combine = combine
+        self.bitmask = bitmask
+        self.converter = converter
+
+        ## End of user defined properties
+        self.shape = [0, 0, 0]
+        self.max_labels = 0
+        self.dtype = np.uint8
+        self.bits = 8
+
+    def get_labels(self, ohifviewer_info):
+        """
+        gather_ROI_info extracts label-name along with bitmasked index and RGBA
+            color code for each distinct label in the ROI collection.
+
+        Args:
+            file_obj (flywheel.models.file.File): The flywheel file-object with
+                ROI data contained within the `.info.ROI` metadata.
+
+        Returns:
+            OrderedDict: the label object populated with ROI attributes
+        """
+
+        # dictionary for labels, index, R, G, B, A
+        labels = OrderedDict()
+
+        # React OHIF Viewer
+        roi_list = [
+            individual_roi
+            for roi_type_list in ohifviewer_info.values()
+            for individual_roi in roi_type_list
+        ]
+
+        for roi in roi_list:
+            if roi.get("location"):
+                if roi["location"] not in labels.keys():
+                    label_index = int(2 ** (len(labels)))
+                    labels[roi["location"]] = RoiLabel(
+                        label=roi["location"],
+                        index=label_index,
+                        color="fbbc05",
+                        RGB=[int("fbbc05"[i: i + 2], 16) for i in [1, 3, 5]],
+                    )
+            else:
+                log.warning(
+                    "There is an ROI without a label. To include this ROI in the "
+                    "output, please attach a label."
+                )
+        return labels
+
+    @abstractmethod
+    def create(self, ohifviewer_info):
+        pass
+
+    @abstractmethod
+    def get_affine(self):
+        pass
+
+
+@dataclass
+class Creator:
+    orig_dir: Path
+    roi_dir: Path
+    combine: bool
+    bitmask: bool
+    output_dir: Path
+    base_file_name: str
+    creator: CreateWorker = None
+    converter: Converters.Converter = None
+    labels: dict = None
+
+
+
+    def __post_init__(self):
+
+        self.creator = self.creator(
+            orig_dir=self.orig_dir,
+            roi_dir=self.roi_dir,
+            output_dir=self.output_dir,
+            base_file_name=self.base_file_name,
+            combine=self.combine,
+            bitmask=self.bitmask,
+            converter=self.converter
+        )
+
+    def create(self, ohifviewer_info):
+        labels = self.creator.create(ohifviewer_info)
+        return labels
+
+    def get_affine(self):
+        return self.creator.get_affine()
+
+
+class DicomCreator(CreateWorker):
+    def __init__(self, orig_dir, roi_dir, output_dir, base_file_name, combine, bitmask, converter):
```

Comment:
SO let me just make sure I'm understanding this here.

* You're finding all ROIs that match this particular dicom by Study/SeriesInstanceUID.
* Then for each ROI you find the SOP and find the corresponding dicom slice
* Then you "Get the ROI data" (This part seems like magic to me, I just don't fully understand)
* Then you save the ROI data as a dicom slice
* Finally you pass all the dicom slices to a converter of choice that converts from dicom to nifti?

(No comment here, just making sure I'm understanding)

Reply:
yes